File: backend/app/schemas/formular.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

# Feldtypen des Formular-Moduls. Dropdowns nutzen `optionen`; sterne/skala nutzen
# `max_sterne` als Maximum; datei = Datei-Upload.
ERLAUBTE_FELDTYPEN = {
    "text",
    "mehrzeilig",
    "checkbox",
    "sterne",
    "skala",
    "dropdown",
    "dropdown_mehrfach",
    "datum",
    "zahl",
    "email",
    "telefon",
    "ja_nein",
    "datei",
}
DROPDOWN_TYPEN = {"dropdown", "dropdown_mehrfach"}


# --- Felder ------------------------------------------------------------------


class FormularFeldOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    label: str
    typ: str
    pflicht: bool
    hinweis: str | None
    optionen: list[str]
    max_sterne: int
    reihenfolge: int
    aktiv: bool


class FormularFeldCreate(BaseModel):
    label: str = Field(min_length=1, max_length=255)
    typ: str = "text"
    pflicht: bool = False
    hinweis: str | None = None
    optionen: list[str] = Field(default_factory=list)
    max_sterne: int = Field(default=5, ge=1, le=10)
    reihenfolge: int = 0
    aktiv: bool = True

    @field_validator("typ")
    @classmethod
    def typ_gueltig(cls, wert: str) -> str:
        if wert not in ERLAUBTE_FELDTYPEN:
            raise ValueError(f"typ muss einer von {sorted(ERLAUBTE_FELDTYPEN)} sein")
        return wert


class FormularFeldUpdate(BaseModel):
    label: str | None = Field(default=None, min_length=1, max_length=255)
    typ: str | None = None
    pflicht: bool | None = None
    hinweis: str | None = None
    optionen: list[str] | None = None
    max_sterne: int | None = Field(default=None, ge=1, le=10)
    reihenfolge: int | None = None
    aktiv: bool | None = None

    @field_validator("typ")
    @classmethod
    def typ_gueltig(cls, wert: str | None) -> str | None:
        if wert is not None and wert not in ERLAUBTE_FELDTYPEN:
            raise ValueError(f"typ muss einer von {sorted(ERLAUBTE_FELDTYPEN)} sein")
        return wert


# --- Formulare ---------------------------------------------------------------


class FormularOut(BaseModel):
    """Admin-Sicht auf ein Formular inkl. aller Felder."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    beschreibung: str | None
    aktiv: bool
    login_erforderlich: bool
    email_empfaenger: str | None
    gruppenfuehrer_sichtbar: bool
    start_am: datetime | None
    ablauf_am: datetime | None
    max_einreichungen: int | None
    aufbewahrung_tage: int | None
    danke_text: str | None
    ergebnis_oeffentlich: bool
    einwilligung_text: str | None
    mehrfach_verhindern: bool
    reihenfolge: int
    felder: list[FormularFeldOut] = []


class FormularCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    beschreibung: str | None = None
    aktiv: bool = False
    login_erforderlich: bool = False
    email_empfaenger: str | None = Field(default=None, max_length=255)
    gruppenfuehrer_sichtbar: bool = False
    start_am: datetime | None = None
    ablauf_am: datetime | None = None
    max_einreichungen: int | None = None
    aufbewahrung_tage: int | None = None
    danke_text: str | None = None
    ergebnis_oeffentlich: bool = False
    einwilligung_text: str | None = None
    mehrfach_verhindern: bool = False
    reihenfolge: int = 0


class FormularUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    beschreibung: str | None = None
    aktiv: bool | None = None
    login_erforderlich: bool | None = None
    email_empfaenger: str | None = Field(default=None, max_length=255)
    gruppenfuehrer_sichtbar: bool | None = None
    # Datum-/Text-Felder: explizit null = löschen; weggelassen = unverändert
    # (`model_dump(exclude_unset=True)` im Service).
    start_am: datetime | None = None
    ablauf_am: datetime | None = None
    max_einreichungen: int | None = None
    aufbewahrung_tage: int | None = None
    danke_text: str | None = None
    ergebnis_oeffentlich: bool | None = None
    einwilligung_text: str | None = None
    mehrfach_verhindern: bool | None = None
    reihenfolge: int | None = None


class FormularOeffentlichOut(BaseModel):
    """Öffentliche Sicht (Kiosk/Mitglied): nur was zum Ausfüllen nötig ist –
    ohne E-Mail-Empfänger und Gruppenführer-Sichtbarkeit."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    beschreibung: str | None
    login_erforderlich: bool
    danke_text: str | None
    einwilligung_text: str | None
    ergebnis_oeffentlich: bool
    felder: list[FormularFeldOut] = []


# --- Einreichungen -----------------------------------------------------------


class EinreichungCreate(BaseModel):
    # Antworten je Feld: {feld_id (als String) -> Wert}. Wert-Typ hängt vom Feldtyp ab
    # (str / bool / int / list[str]); die inhaltliche Validierung macht der Service.
    antworten: dict[str, object] = Field(default_factory=dict)
    # DSGVO-Einwilligung (nur nötig, wenn das Formular einen Einwilligungstext hat).
    einwilligung: bool = False
    # Honeypot gegen Bots: von echten Nutzern nie ausgefüllt (verstecktes Feld).
    hp: str = ""


class EinreichungAntwortOut(BaseModel):
    feld_id: int
    label: str
    typ: str
    wert: object


class EinreichungOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    formular_id: int
    person_id: int | None
    person_name: str | None
    antworten: list[EinreichungAntwortOut]
    erstellt_am: datetime


# --- Zusammenfassung / Auswertung --------------------------------------------


class FeldZusammenfassung(BaseModel):
    feld_id: int
    label: str
    typ: str
    anzahl_beantwortet: int
    # Ø bei Sternebewertung
    durchschnitt: float | None = None
    # Anzahl je Ausprägung (Sterne, Dropdown-Optionen, Ja/Nein)
    verteilung: dict[str, int] | None = None
    # Einzelantworten bei Freitextfeldern
    texte: list[str] | None = None


class ZusammenfassungOut(BaseModel):
    formular_id: int
    name: str
    anzahl_einreichungen: int
    ablauf_am: datetime | None
    felder: list[FeldZusammenfassung]

File: frontend/src/api/formular.ts
import { apiDelete, apiGet, apiPost, apiPut, apiUpload } from "./client";

export type FormularFeldTyp =
  | "text"
  | "mehrzeilig"
  | "checkbox"
  | "sterne"
  | "skala"
  | "dropdown"
  | "dropdown_mehrfach"
  | "datum"
  | "zahl"
  | "email"
  | "telefon"
  | "ja_nein"
  | "datei";

export interface FormularFeld {
  id: number;
  label: string;
  typ: FormularFeldTyp;
  pflicht: boolean;
  hinweis: string | null;
  optionen: string[];
  max_sterne: number;
  reihenfolge: number;
  aktiv: boolean;
}

export interface Formular {
  id: number;
  name: string;
  beschreibung: string | null;
  aktiv: boolean;
  login_erforderlich: boolean;
  email_empfaenger: string | null;
  moderator_sichtbar: boolean;
  start_am: string | null;
  ablauf_am: string | null;
  max_einreichungen: number | null;
  aufbewahrung_tage: number | null;
  danke_text: string | null;
  ergebnis_oeffentlich: boolean;
  einwilligung_text: string | null;
  mehrfach_verhindern: boolean;
  reihenfolge: number;
  felder: FormularFeld[];
}

export interface FeldZusammenfassung {
  feld_id: number;
  label: string;
  typ: FormularFeldTyp;
  anzahl_beantwortet: number;
  durchschnitt: number | null;
  verteilung: Record<string, number> | null;
  texte: string[] | null;
}

export interface Zusammenfassung {
  formular_id: number;
  name: string;
  anzahl_einreichungen: number;
  ablauf_am: string | null;
  felder: FeldZusammenfassung[];
}

export interface FormularOeffentlich {
  id: number;
  name: string;
  beschreibung: string | null;
  login_erforderlich: boolean;
  danke_text: string | null;
  einwilligung_text: string | null;
  ergebnis_oeffentlich: boolean;
  felder: FormularFeld[];
}

export interface EinreichungAntwort {
  feld_id: number;
  label: string;
  typ: FormularFeldTyp;
  wert: unknown;
}

export interface Einreichung {
  id: number;
  formular_id: number;
  person_id: number | null;
  person_name: string | null;
  antworten: EinreichungAntwort[];
  erstellt_am: string;
}

export type FormularEingabe = Partial<Omit<Formular, "id" | "felder">>;
export type FeldEingabe = Partial<Omit<FormularFeld, "id">> & { label: string; typ: FormularFeldTyp };

// --- Admin -------------------------------------------------------------------

export const holeFormulare = () => apiGet<Formular[]>("/gruppenfuehrer/formulare");
export const holeFormular = (id: number) => apiGet<Formular>(`/gruppenfuehrer/formulare/${id}`);
export const formularAnlegen = (daten: { name: string }) =>
  apiPost<Formular>("/gruppenfuehrer/formulare", daten);
export const formularAktualisieren = (id: number, daten: FormularEingabe) =>
  apiPut<Formular>(`/gruppenfuehrer/formulare/${id}`, daten);
export const formularLoeschen = (id: number) => apiDelete<void>(`/gruppenfuehrer/formulare/${id}`);

export const feldAnlegen = (formularId: number, daten: FeldEingabe) =>
  apiPost<FormularFeld>(`/gruppenfuehrer/formulare/${formularId}/felder`, daten);
export const feldAktualisieren = (feldId: number, daten: FeldEingabe) =>
  apiPut<FormularFeld>(`/gruppenfuehrer/formulare/felder/${feldId}`, daten);
export const feldLoeschen = (feldId: number) =>
  apiDelete<void>(`/gruppenfuehrer/formulare/felder/${feldId}`);

export const holeEinreichungen = (formularId: number) =>
  apiGet<Einreichung[]>(`/gruppenfuehrer/formulare/${formularId}/einreichungen`);
export const holeZusammenfassung = (formularId: number) =>
  apiGet<Zusammenfassung>(`/gruppenfuehrer/formulare/${formularId}/zusammenfassung`);
export const holeSichtbareFormulare = () =>
  apiGet<Formular[]>("/gruppenfuehrer/formulare/sichtbar");
export const formularDuplizieren = (id: number) =>
  apiPost<Formular>(`/gruppenfuehrer/formulare/${id}/duplizieren`);
/** CSV-Export herunterladen. Über einen authentifizierten Blob-Request (nicht als
 * <a href>, denn ein Link kann den Bearer-Token nicht mitsenden → 401). */
export async function formularCsvHerunterladen(id: number, name: string): Promise<void> {
  const blob = await apiGet<Blob>(`/gruppenfuehrer/formulare/${id}/export.csv`);
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = `formular-${name.replace(/[^\w.-]+/g, "_") || id}.csv`;
  document.body.appendChild(a);
  a.click();
  a.remove();
  URL.revokeObjectURL(url);
}

// --- Öffentlich / Mitglied ---------------------------------------------------

export const holeOeffentlicheFormulare = () => apiGet<FormularOeffentlich[]>("/formulare");
export const holeOeffentlichesFormular = (id: number) =>
  apiGet<FormularOeffentlich>(`/formulare/${id}`);
export const holeOeffentlichesErgebnis = (id: number) =>
  apiGet<Zusammenfassung>(`/formulare/${id}/ergebnis`);
export const formularDateiHochladen = (id: number, datei: File) =>
  apiUpload<{ referenz: string }>(`/formulare/${id}/datei`, datei);
export const formularEinreichen = (
  id: number,
  antworten: Record<string, unknown>,
  extra?: { einwilligung?: boolean; hp?: string }
) =>
  apiPost<{ ok: boolean; einreichung_id: number }>(`/formulare/${id}/einreichen`, {
    antworten,
    einwilligung: extra?.einwilligung ?? false,
    hp: extra?.hp ?? "",
  });
